### DISCUSSION TITLE
Help with chunk response

### DISCUSSION THREAD

**Discussion by LorenzoCarretta** [cesanta/mongoose#2082] [Ask questions, get answers] [answered]
Hello everyone!
I'm using version 7.8 and I switched from version 7.6 few month ago. During some regression tests I now have some new issue with response to chunked messages.
I use chunk encoding to pass some files to my device, and it works just fine when I handle the file and return status 200.
But if I need to return an error message (for example I used to check for file size before deciding if I should allow the transfer) then when I try to return the 500 or 400 error using **mg_http_reply**, I see the connection reset and a fetch failed on frontend.
So I wanted to ask you what's the correct way to return an error message to frontend when dealing with a chunk message, without receiving the whole message? I need to stop the transfer right away returning a message to do error handling on frontend.
As I already said in version 7.6 it worked fine just using **mg_http_reply** and then closing the connection, but now it doesn't work anymore.
Thanks!

**Comment by scaprile** [comment, marked as answer]
> I'm using version 7.8 and I switched from version 7.6 few month ago.

I suggest you use 7.9, as chunk processing has changed again

> But if I need to return an error message (for example I used to check for file size before deciding if I should allow the transfer) then when I try to return the 500 or 400 error using **mg_http_reply**, I see the connection reset and a fetch failed on frontend.

Well, I'd need to see a network sniff capture file to understand what you are saying, and the relevant part of the Mongoose event handler to understand what you are doing (and not doing).

I suggest you check what we do and how we do it first, and then, if that does not work for you, open an issue filling the template fields and giving us all information to reproduce your problem.

https://mongoose.ws/tutorials/file-uploads/
https://github.com/cesanta/mongoose/tree/master/examples/file-upload-html-form
https://github.com/cesanta/mongoose/tree/master/examples/file-upload-multiple-posts
https://github.com/cesanta/mongoose/tree/master/examples/file-upload-single-post

**Reply by LorenzoCarretta** [reply]
Thank you, 
![immagine](https://user-images.githubusercontent.com/44646861/220582520-c0d877b5-066f-4e85-8023-40f82ee6319e.png)
Here is the sniff, as you see I send the chunks of a file (14MB) to my device, where we have a 5MB max size. 
First thing I check the size as hm->body.len in my application, and since is bigger than 5MB I send a 413 error. It's sent multiple times (I supposed for each chunk hanlded) but by frontend can't handle it correctly, so it'll reset the connection.
As I said, same code used to work in 7.6

So my question is, what is the best way to decline the file transfer from my device in this case?
Thank you

**Reply by scaprile** [reply]
`mg_http_reply(c, 413, "", "Can't fit that\n"), c->is_draining=1;` works for me
```
$ curl -H Expect: -H "Transfer-Encoding: chunked" --data-binary @filename http://127.0.0.1:8000/upload?name=myfile
Can't fit that
```

Again
> I'd need to see a network sniff capture **file** to understand what you are saying, and **the relevant part of the Mongoose event handler** to understand what you are doing (and not doing).

File, not snapshot; and your event handler code, minimum possible reproducing your issue
```
static int i = 0;
static void cb(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
  struct mg_http_message *hm = (struct mg_http_message *) ev_data;
  if (ev == MG_EV_HTTP_CHUNK && mg_http_match_uri(hm, "/upload")) {
    mg_http_delete_chunk(c, hm);
    if(++i >= 3) mg_http_reply(c, 413, "", "Can't fit that\n"), c->is_draining=1;
    if (hm->chunk.len == 0) {
      mg_http_reply(c, 200, "", "ok (chunked)\n");
    }
```
> If that does not work for you, **open an issue filling the template fields and giving us all information to reproduce your problem**.

**Reply by LorenzoCarretta** [reply]
Hello,
sorry I misunderstood, here are the files you asked.
[snippets.zip](https://github.com/cesanta/mongoose/files/10823122/snippets.zip)

As you see I tried using is_draining but unsuccessfully

**Reply by scaprile** [reply]
No, I beg to differ, I don't see that.
Leaving aside the fact that your code snippet is not consistent with the network capture, and assuming you are doing something like this instead:
```
    if (ev == MG_EV_HTTP_MSG)  {
        if (mg_http_match_uri(hm, "/upload")) do some customer function
        /* ... various else if not important */
    } else if (ev == MG_EV_HTTP_CHUNK && mg_http_match_uri(hm, "/upload")) {
        do the same customer function
    }
```
Then, inside your function, you do this:
```
   if (hm->body.len > CERTIFICATE_MAX_SIZE)
    {
        mg_http_delete_chunk(c, hm);
        **mg_http_reply(c, 413, "", "operation running");
        c->is_draining = 1;**
        return ;
    }
```
I don't see '413' elsewhere, so this should be the point where you say you do what was suggested.
1. In the examples and tutorials linked on the first response, we never check for `hm->body.len` inside an `MG_EV_HTTP_CHUNK` event, that element is only initialized after the last chunk has been received. If you need a "so far" byte count, you'll have to do it yourself.
2. Wireshark's conversation statistics window shows 4 TCP conversations, 4 different TCP port numbers, 3 of them are 10KB data transfers ended with a 413 code. I only get to see the fourth connection establishment, but it is clear to me that connections are being closed:
> Frame 15 for port 52533
> Frame 30 for port 52535
> Frame 38 for port 53219

Whatever your client is trying to do, it seems to be retrying the transfer and Mongoose is closing the connection with 413

At this point I'm closing this question. If after fixing your issues you need further help, please open an issue filling the template fields and giving us all information to reproduce your problem.
